tensorflow/python/data/experimental/kernel_tests/override_threadpool_test.py

```diff
@@ -70,17 +69,15 @@ def get_thread_id(_):
       # perform work.
       self.assertLessEqual(len(thread_ids), num_threads)
 
-  @parameterized.named_parameters(
-      ("1", 1, None),
-      ("2", 2, None),
-      ("3", 4, None),
-      ("4", 8, None),
-      ("5", 16, None),
-      ("6", 4, -1),
-      ("7", 4, 0),
-      ("8", 4, 1),
-      ("9", 4, 4),
-  )
+  @combinations.generate(
+      combinations.times(
+          test_base.default_test_combinations(),
+          combinations.combine(
+              num_threads=[1, 2, 4, 8, 16], max_intra_op_parallelism=[None]) +
+          combinations.combine(
+              num_threads=[4], max_intra_op_parallelism=[0, 1, 4]) +
+          combinations.combine(
```

Comment:
are you adding this as a new combination?

Reply:
Yes, I made a little change here (changing {`num_threads=4` & `max_intra_op_parallelism=-1`} to {`num_threads=5` & `max_intra_op_parallelism=-1`} ). 

The reason is that if using `combinations.combine(num_threads=[4], max_intra_op_parallelism=[0, 1, 4, -1])`(which will be exactly same as the original test), there will be name conflicts in the generated test functions between the test case {`num_threads=4` & `max_intra_op_parallelism=-1`}  and the test case {`num_threads=4` & `max_intra_op_parallelism=1`}, becase the decorator will convert `-1` to `1` when combining the parameter values into the function name.  

So I made a little change by adding this new combination where `num_threads` is changed from `4` to `5` when `max_intra_op_parallelism=-1`.

Reply:
What if you replace `-1` with `None`?

Reply:
@jsimsa According to `bazel-out/darwin-opt/bin/tensorflow/python/ops/gen_experimental_dataset_ops.py Line#7322-7323`, `None` means `max_intra_op_parallelism = 1`, so I think it will be different. Also {`num_threads=4`, `max_intra_op_parallelism=None`} has been covered by [the first combination](https://github.com/tensorflow/tensorflow/blob/166b4bb3a35f8e8ff5d58e2558918e7258b6d5f6/tensorflow/python/data/experimental/kernel_tests/override_threadpool_test.py#L75-L76).

```
def thread_pool_handle(num_threads, display_name, max_intra_op_parallelism=1, container="", shared_name="", name=None):
  r"""Creates a dataset that uses a custom thread pool to compute `input_dataset`.

  Args:
    num_threads: An `int`. The number of threads in the thread pool.
    display_name: A `string`.
      A human-readable name for the threads that may be visible in some
      visualizations.
      threadpool.
    max_intra_op_parallelism: An optional `int`. Defaults to `1`.
      The maximum degree of parallelism to use within operations that execute on this
      threadpool.
    container: An optional `string`. Defaults to `""`.
    shared_name: An optional `string`. Defaults to `""`.
    name: A name for the operation (optional).

  Returns:
    A `Tensor` of type `resource`.
  """
  _ctx = _context._context or _context.context()
  tld = _ctx._thread_local_data
  if tld.is_eager:
    try:
      _result = _pywrap_tensorflow.TFE_Py_FastPathExecute(
        _ctx._context_handle, tld.device_name, "ThreadPoolHandle", name,
        tld.op_callbacks, "num_threads", num_threads,
        "max_intra_op_parallelism", max_intra_op_parallelism, "display_name",
        display_name, "container", container, "shared_name", shared_name)
      return _result
    except _core._FallbackException:
      try:
        return thread_pool_handle_eager_fallback(
            num_threads=num_threads,
            max_intra_op_parallelism=max_intra_op_parallelism,
            display_name=display_name, container=container,
            shared_name=shared_name, name=name, ctx=_ctx)
      except _core._SymbolicException:
        pass  # Add nodes to the TensorFlow graph.
    except _core._NotOkStatusException as e:
      _ops.raise_from_not_ok_status(e, name)
  # Add nodes to the TensorFlow graph.
  num_threads = _execute.make_int(num_threads, "num_threads")
  display_name = _execute.make_str(display_name, "display_name")
  if max_intra_op_parallelism is None:
    max_intra_op_parallelism = 1
  max_intra_op_parallelism = _execute.make_int(max_intra_op_parallelism, "max_intra_op_parallelism")
  if container is None:
    container = ""
  container = _execute.make_str(container, "container")
  if shared_name is None:
    shared_name = ""
  shared_name = _execute.make_str(shared_name, "shared_name")
  _, _, _op, _outputs = _op_def_library._apply_op_helper(
        "ThreadPoolHandle", num_threads=num_threads,
                            display_name=display_name,
                            max_intra_op_parallelism=max_intra_op_parallelism,
                            container=container, shared_name=shared_name,
                            name=name)
  _result = _outputs[:]
  if _execute.must_record_gradient():
    _attrs = ("num_threads", _op._get_attr_int("num_threads"),
              "max_intra_op_parallelism",
              _op._get_attr_int("max_intra_op_parallelism"), "display_name",
              _op.get_attr("display_name"), "container",
              _op.get_attr("container"), "shared_name",
              _op.get_attr("shared_name"))
    _inputs_flat = _op.inputs
    _execute.record_gradient(
        "ThreadPoolHandle", _inputs_flat, _attrs, _result)
  _result, = _result
  return _result
```

Reply:
I see. In that case, use the following pattern:
```
combinations.combine(
              num_threads=[4], max_intra_op_parallelism=[0, 1, 4]) +
combinations.combine(
              num_threads=[4], max_intra_op_parallelism=combinations.NamedObject("MinusOne", -1))
```

The `combinations.NamedObject` allows you to decide what name should be used for the particular combination.

Reply:
It looks like `combinations.NamedObject` requires the second parameter to be a function so that it can be callable [here](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/framework/test_combinations.py#L400-L401).

Reply:
That means it needs to be 
```
combinations.combine(
    num_threads=[4],
    max_intra_op_parallelism=combinations.NamedObject("MinusOne", lambda: -1))
```

Reply:
How about this:
```Python
  @combinations.generate(
      combinations.times(
          test_base.default_test_combinations(),
          combinations.combine(
              num_threads=[1, 2, 4, 8, 16], max_intra_op_parallelism=[None]) +
          combinations.combine(
              num_threads=[4], max_intra_op_parallelism=[0, 1, 4]) +
          combinations.combine(
              num_threads=[4],
              max_intra_op_parallelism=combinations.NamedObject("MinusOne", -1))
      ))
  def testNumThreadsDeprecated(self, num_threads, max_intra_op_parallelism):
    if isinstance(max_intra_op_parallelism, combinations.NamedObject):
      max_intra_op_parallelism = max_intra_op_parallelism._obj

    def override_threadpool_fn(dataset):
      return threadpool.override_threadpool(
          dataset,
          threadpool.PrivateThreadPool(
              num_threads,
              max_intra_op_parallelism=max_intra_op_parallelism,
              display_name="private_thread_pool_%d" % num_threads))

    self._testNumThreadsHelper(num_threads, override_threadpool_fn)
```

Reply:
No, that's not great because it relies on implementation details of `NamedObject`. But since max_intra_op_parallelism=None implies 1, `combinations.combine(num_threads=[1, 2, 4, 8, 16], max_intra_op_parallelism=[None])` covers the case of `num_threads=4` and `max_intra_op_parallelism=None` so we can simply use `combinations.combine(num_threads=[4], max_intra_op_parallelism=[-1, 0, 4])` as the second (and last combination). Makes sense?

Reply:
Yeah, sounds good! I will change it to your suggestion.